Add more perf-tests, new features to FA

IzzyPutterman opened the pull request:
Adding new tests across the board for float32, bfloat16, non-powers-of-2 shapes (to test masks), and tests on sequence parallel for atomics.  This also adds the sequence parallel features from https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py.
I am not sure about the best way to grab the baseline benchmarking numbers.  I have access to V100s and A100s, but I saw on the tests it mentions " # A100 in the CI server is slow-ish for some reason.
    # On some other servers, we are getting about 90% peak for 8kx8x8k float16".  Current plan is to run CI here and use those numbers for baseline, then match against my GPUs as a sanity check.

Jokeren commented:
Thanks for the modification. There was a discussion about using atomic_add a few months ago. It's generally good to achieve higher performance, but I'm not sure if we want the non-deterministic integrated into main. Let's hold this PR for now until Phil is back.

IzzyPutterman commented:
That makes sense, we have a separate version that drops the atomic_adds for separate accumulators and a sum at the end.  I just pushed this version, let me know if it looks better.

tridao commented:
@IzzyPutterman might be good add a comment to refer to the implementation in HazyResearch (e.g. "parallelizing the backward on the seqlen dimension as done in https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py").

IzzyPutterman commented:
> @IzzyPutterman might be good add a comment to refer to the implementation in HazyResearch (e.g. "parallelizing the backward on the seqlen dimension as done in https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py").

Great point, added a snippet with a link at the top.

ptillet commented:
This looks good to me. I do expect perf numbers to fail on our CI server, but this should be a relatively easy fix

IzzyPutterman commented:
Added the changes from the faster-attention branch.  I made some slight changes to the dtypes, which were set to fp16 in the forward.  I changed them to be the dtype of K, V so that BF16 also should work.  

The perf regression tests default to causal True and I made a small change to the tutorial to speed up the reference pytorch implementation for the masking.

janEbert commented:
Hey! Thanks @IzzyPutterman and @ptillet for the great new implementation! :)

Most of the numbers look absolutely superb, even surpassing the FlashAttention reference kernel in the backward pass. However, just FYI, I'm getting about 60 % the performance of the FlashAttention reference kernel on a 40GB A100 in the causal forward pass with either sequence parallelism setting (activated or deactivated).

```
fused-attention-batch4-head48-d64-fwd-causal:
     N_CTX      Triton      Flash
0   1024.0   90.799438  48.967322
1   2048.0  114.020909  70.995286
2   4096.0  131.016547  82.205899
3   8192.0  143.605973  89.178259
4  16384.0  150.926554  92.484222
```

With the sequence-parallel version I get OOMs in the backward pass when using a head dimension of 64, so I turned that down to 32.

```
fused-attention-batch4-head48-d32-fwd-causal-seqparallel:
     N_CTX      Triton      Flash
0   1024.0   68.793519  38.192535
1   2048.0   84.023941  53.051927
2   4096.0   78.063818  50.116476
3   8192.0  108.115386  62.992597
4  16384.0  113.435417  64.512247
```

<details>
<summary>Full benchmark results</summary>
Note that I simply adjusted the tutorial benchmark function to use the version in `python/triton/ops/flash_attention.py`.

<pre>
fused-attention-batch4-head48-d64-fwd:
     N_CTX      Triton       Flash
0   1024.0  100.695247   87.931126
1   2048.0  142.757964  142.069599
2   4096.0  149.949147  165.684746
3   8192.0  152.291195  178.378226
4  16384.0  152.972838  186.020589
fused-attention-batch4-head48-d64-fwd-causal:
     N_CTX      Triton      Flash
0   1024.0   90.799438  48.967322
1   2048.0  114.020909  70.995286
2   4096.0  131.016547  82.205899
3   8192.0  143.605973  89.178259
4  16384.0  150.926554  92.484222
fused-attention-batch4-head48-d64-bwd:
     N_CTX     Triton       Flash
0   1024.0  72.925357  110.173990
1   2048.0  85.362130  136.248001
2   4096.0  92.770819  160.109770
3   8192.0  94.962014  174.743706
4  16384.0  97.053800  182.725176
fused-attention-batch4-head48-d64-bwd-causal:
     N_CTX     Triton      Flash
0   1024.0  51.941537  55.078529
1   2048.0  66.542515  68.068726
2   4096.0  77.177173  80.050337
3   8192.0  83.898314  87.356685
4  16384.0  87.580424  91.382803

fused-attention-batch4-head48-d32-fwd-seqparallel:
     N_CTX      Triton       Flash
0   1024.0   73.445880   73.000930
1   2048.0  104.645858  105.963907
2   4096.0  110.380061  122.839028
3   8192.0  112.607945  125.972521
4  16384.0  113.683626  128.993496
fused-attention-batch4-head48-d32-fwd-causal-seqparallel:
     N_CTX      Triton      Flash
0   1024.0   68.793519  38.192535
1   2048.0   84.023941  53.051927
2   4096.0   78.063818  50.116476
3   8192.0  108.115386  62.992597
4  16384.0  113.435417  64.512247
fused-attention-batch4-head48-d32-bwd-seqparallel:
     N_CTX     Triton       Flash
0   1024.0  39.983359   75.358061
1   2048.0  53.937941  102.604451
2   4096.0  57.003315  117.601057
3   8192.0  57.689181  126.746714
4  16384.0  57.567115  131.653772
fused-attention-batch4-head48-d32-bwd-causal-seqparallel:
     N_CTX     Triton      Flash
0   1024.0  37.227551  44.760162
1   2048.0  43.555031  51.307885
2   4096.0  47.301705  58.891837
3   8192.0  48.826570  63.368705
4  16384.0  48.885554  65.820965

</pre>
</details>

IzzyPutterman commented:
Hey,
Good to hear about the backward pass!  Sequence Parallel is only an option for the backward and has no effect on the forward.  The forward is taken from this PR which switches to block_ptrs and faster math: https://github.com/openai/triton/pull/1917

janEbert commented:
You're absolutely right, I actually realized that but forgot to remove sequence parallelism from the forward benchmarks.
Thanks!

ptillet commented:
```
fused-attention-batch4-head48-d64-fwd:
     N_CTX      Triton       Flash
0   1024.0  100.695247   87.931126
1   2048.0  142.757964  142.069599
2   4096.0  149.949147  165.684746
3   8192.0  152.291195  178.378226
4  16384.0  152.972838  186.020589
```
There's a bug here https://github.com/openai/triton/blob/main/python/tutorials/06-fused-attention.py#L412, sorry. I think the perf number that you're seeing just comes from the fact that flash attention always use causal but multiplies flops count by two for non causal, so the numbers are 2x inflated.

ptillet commented:
should be fixed now; you can retry :)
